### PR TITLE
Tool for registering multiple foreign participants

### PR DESCRIPTION
**Pull request by bergie**
Refs #74

### CODE REVIEW

File: src/utils/register-foreign-multiple.coffee [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  [role, definition] = val.split ':'
+  map[role] = definition
+  return map
+
+main = ->
+  program
+    .option('--broker <uri>', 'Broker address', String, '')
+    .option('--roles <role:definition>', 'Map of roles and definition files', collectMap, {})
+    .option('--interval <SECONDS>', 'How often to send discovery message', Number, 60)
+    .option('--forever <true>', 'Keep running forever', Boolean, false)
```
Comment (jonnor) [inline src/utils/register-foreign-multiple.coffee]: `--forever` should not be an option I think, but instead always on. We generally require messages to be sent periodically now (#30)

Comment (bergie) [inline src/utils/register-foreign-multiple.coffee]: OK, switched to on by default

File: src/coordinator.coffee [hunk 2]
```diff
@@ -212,6 +212,9 @@ class Coordinator extends EventEmitter
     return @library.getSource component, callback
 
   startParticipant: (node, component, callback) ->
+    if @options.ignore?.length and node in @options.ignore
```
Comment (jonnor) [inline src/coordinator.coffee]: Unrelated, but ok.

File: src/utils/register-foreign-multiple.coffee [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  Promise.map Object.keys(program.roles), (role) ->
+    defPath = path.resolve process.cwd(), program.roles[role]
+    readFile defPath, 'utf-8'
+    .then (contents) ->
+      return Promise.reject "No definition found in #{defPath}" unless contents
+      definition = yaml.safeLoad contents
+      return Promise.reject new Error ".component is not defined" unless definition.component
+      definition.role = role
+      definition.id = definition.role unless definition.id
+      definition = foreigner.mapPorts definition
```
Comment (jonnor) [inline src/utils/register-foreign-multiple.coffee]: Where is `#ROLE` and such being replaced for queues? I don't see this anywere.
Needed to be compatible with `msgflo-register-foreign`, and use with .yaml files declaring a component which may have multiple instances.

File: src/utils/register-foreign-multiple.coffee [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+          Promise.resolve definition
+  .then (definitions) ->
+    onComplete() unless program.forever
+    console.log '--forever enabled'
+    setInterval ->
+      Promise.map definitions, (definition) ->
+        register messaging, definition
+      .asCallback (err) ->
+        console.log 'Warning: Failed to send discovery message:', err if err
+    , program.interval*1000/2.2
```
Comment (jonnor) [inline src/utils/register-foreign-multiple.coffee]: This new tool needs at least one test checking that we do indeed get valid discovery messages out when started.

File: src/utils/register-foreign-multiple.coffee (hunk 3 above)
Comment (bergie) [inline src/utils/register-foreign-multiple.coffee]: That is what `foreigner.mapPorts` does for you

File: src/utils/register-foreign-multiple.coffee [hunk 5]
```diff
@@ -0,0 +1,65 @@
+program = require 'commander'
+msgflo_nodejs = require 'msgflo-nodejs'
+fs = require 'fs'
+path = require 'path'
+foreigner = require '../foreign-participant'
```
Comment (jonnor) [inline src/utils/register-foreign-multiple.coffee]: Noe that this is not really needed to have a `Participant` instance just to send a discovery message. Can also us `msgfloNodejs.Client.registerParticipant(definition)`

Comment (bergie) [inline src/utils/register-foreign-multiple.coffee]: Yes, but the `foreign-participant` lib provides the `#ROLE` mapping magic we need here

**Pull request merged**